File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/routing/DelegatingRoutingHandler.java
package uk.nhs.nhsx.core.routing;

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent;

import static uk.nhs.nhsx.core.routing.Routing.Match.Matched;
import static uk.nhs.nhsx.core.routing.Routing.RouterMatch.matched;

public abstract class DelegatingRoutingHandler implements Routing.RoutingHandler {
    private final Routing.RoutingHandler delegate;

    protected DelegatingRoutingHandler(Routing.RoutingHandler delegate) {
        this.delegate = delegate;
    }

    public Routing.RoutingHandler getDelegate() {
        return delegate;
    }

    @Override
    public final Routing.RouterMatch match(APIGatewayProxyRequestEvent request) {
        Routing.RouterMatch match = delegate.match(request);
        return match.matchType == Matched ? matched(this) : match;
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/core/routing/DelegatingRoutingHandler.kt
package uk.nhs.nhsx.core.routing

import com.amazonaws.services.lambda.runtime.events.APIGatewayProxyRequestEvent
import uk.nhs.nhsx.core.routing.Routing.RouterMatch.Matched
import uk.nhs.nhsx.core.routing.Routing.RoutingHandler

abstract class DelegatingRoutingHandler protected constructor(val delegate: RoutingHandler) : RoutingHandler {
    override fun match(request: APIGatewayProxyRequestEvent) = when(val match = delegate.match(request)) {
        is Matched -> Matched(this)
        else -> match
    }
}
